internal/sqldb/writer.go

```diff
@@ -267,12 +270,33 @@ func writeObservations(
 	sqlClient *sql.DB,
 	observations []*observation,
 ) error {
+	// TODO: abstract out the batch write logic into a helper function for both
+	// observations and triples (and future) tables.
+	valueStrings := make([]string, 0, batchSize)
+	valueArgs := make([]interface{}, 0, batchSize*5) // 5 columns to insert
 	for _, o := range observations {
-		sqlStmt := `INSERT INTO observations(entity,variable,date,value,provenance) VALUES (?, ?, ?, ?, ?)`
-		_, err := sqlClient.Exec(sqlStmt, o.entity, o.variable, o.date, o.value, o.provenance)
+		sqlValues := "(?, ?, ?, ?, ?)"
+		valueStrings = append(valueStrings, sqlValues)
+		valueArgs = append(valueArgs, o.entity, o.variable, o.date, o.value, o.provenance)
+		if len(valueStrings) >= batchSize {
+			sqlStmt := fmt.Sprintf(`INSERT INTO observations(entity,variable,date,value,provenance) VALUES %s`, strings.Join(valueStrings, ","))
+			_, err := sqlClient.Exec(sqlStmt, valueArgs...)
+			if err != nil {
+				return err
+			}
+			log.Printf("[INSERT] observations: %d entries", batchSize)
+			valueStrings = valueStrings[:0] // reset the slices for the next batch
+			valueArgs = valueArgs[:0]
+		}
+	}
+	// Insert remaining data that didn't make a full batch
+	if len(valueStrings) > 0 {
+		sqlStmt := fmt.Sprintf(`INSERT INTO observations(entity,variable,date,value,provenance) VALUES %s`, strings.Join(valueStrings, ","))
```

Comment:
Consider extracting this into a helper function to avoid duplication of code.

Reply:
Done